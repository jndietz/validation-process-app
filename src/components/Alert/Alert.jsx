import React from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';

import AlertModule from './Alert.module.css';

export const severityLevel = {
    warning: 'warning',
    error: 'error',
    info: 'info',
    success: 'success',
};

const AlertProps = {
    className: PropTypes.string,
    severity: PropTypes.oneOf(Object.keys(severityLevel)),
    message: PropTypes.node.isRequired,
};

/**
 * @typedef {React.HTMLAttributes<?> & PropTypes.InferProps<AlertProps>} AlertPropTypes
 */

/**
 * @type {React.FC<AlertPropTypes>}
 */
export default function Alert({ severity, message, className, ...restProps }) {
    return (
        <div
            {...restProps}
            className={clsx(
                AlertModule.alert,
                AlertModule[`alert-${severity}`],
                className,
            )}
        >
            {message}
        </div>
    );
}

Alert.propTypes = AlertProps;

Alert.defaultProps = {
    severity: 'info',
};
