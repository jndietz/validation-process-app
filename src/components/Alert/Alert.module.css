.alert {
    @apply w-full rounded-md px-4 py-2 border;
}

.alert-success {
    @apply bg-green-100 text-green-700 border-green-400;
}

.alert-info {
    @apply bg-blue-100 text-blue-700 border-blue-400;
}

.alert-warning {
    @apply bg-yellow-100 text-yellow-700 border-yellow-400;
}

.alert-error {
    @apply bg-red-100 text-red-700 border-red-400;
}
