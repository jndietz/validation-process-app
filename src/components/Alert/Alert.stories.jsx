import Alert from './Alert';

export default {
    title: 'Components/Alert',
    component: Alert,
    args: {
        severity: 'info',
        message: 'This is an info alert',
    },
};

const Template = (args) => <Alert {...args} />;

export const Info = Template.bind({});
Info.args = {
    severity: 'info',
    message: 'This is an info alert',
};

export const Success = Template.bind({});
Success.args = {
    severity: 'success',
    message: 'This is a success alert',
};

export const Warning = Template.bind({});
Warning.args = {
    severity: 'warning',
    message: 'This is a warning alert',
};

export const Error = Template.bind({});
Error.args = {
    severity: 'error',
    message: 'This is an error alert',
};
