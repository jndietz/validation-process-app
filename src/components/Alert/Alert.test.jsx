import Alert, { severityLevel } from './Alert';
import { render, screen } from '@testing-library/react';

describe('Alert', () => {
    const unrollBySeverity = Object.values(severityLevel).map((x) => [x]);

    it.each(unrollBySeverity)(
        'Severity level %s matches snapshot',
        (severity) => {
            render(
                <Alert
                    severity={severity}
                    message="Test"
                    data-testid={severity}
                />,
            );

            expect(screen.getByTestId(severity)).toMatchSnapshot();
        },
    );

    it('Default severity level matches snapshot', () => {
        render(<Alert message="Default" />);

        expect(screen.getByText(/Default/)).toMatchSnapshot();
    });

    it('Passes className to child and matches snapshot', () => {
        render(<Alert message="Classy" className="test" />);

        const alert = screen.getByText(/Classy/);

        expect(alert).toMatchSnapshot();
        expect(alert.classList.contains('test')).toBeTruthy();
    });
});
