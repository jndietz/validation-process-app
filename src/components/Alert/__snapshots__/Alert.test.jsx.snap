// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Alert > Default severity level matches snapshot 1`] = `
<div
  class="_alert_6cfd7b _alert-info_6cfd7b"
>
  Default
</div>
`;

exports[`Alert > Passes className to child and matches snapshot 1`] = `
<div
  class="_alert_6cfd7b _alert-info_6cfd7b test"
>
  Classy
</div>
`;

exports[`Alert > Severity level error matches snapshot 1`] = `
<div
  class="_alert_6cfd7b _alert-error_6cfd7b"
  data-testid="error"
>
  Test
</div>
`;

exports[`Alert > Severity level info matches snapshot 1`] = `
<div
  class="_alert_6cfd7b _alert-info_6cfd7b"
  data-testid="info"
>
  Test
</div>
`;

exports[`Alert > Severity level success matches snapshot 1`] = `
<div
  class="_alert_6cfd7b _alert-success_6cfd7b"
  data-testid="success"
>
  Test
</div>
`;

exports[`Alert > Severity level warning matches snapshot 1`] = `
<div
  class="_alert_6cfd7b _alert-warning_6cfd7b"
  data-testid="warning"
>
  Test
</div>
`;
